# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/systemservices/video_engine/video_capture_factory.h
# Commit: fef894fcc429
# Full Hash: fef894fcc4294027fdfb62af3997b87411bef554
# Author: Jan Grulich <jgrulich@redhat.com>
# Date: 2023-11-30 17:23:44
# Regressor Bug: 1844020
# File Overlap Count: 1
# Description:
#   Bug 1844020 - Add a camera device placeholder in case PipeWire is not initialized yet r=pehrsons
#   
#   Implement DeviceInfoPipeWire with option to add a device placeholder
#   when there is a camera device, but we don't have access to PipeWire yet.
#   We ask the Camera portal whether there is a camera device present.
# ==============================================================================

diff -r ed31b2acb5fb -r fef894fcc429 dom/media/systemservices/video_engine/video_capture_factory.h
--- a/dom/media/systemservices/video_engine/video_capture_factory.h	Thu Nov 30 11:49:35 2023 +0000
+++ b/dom/media/systemservices/video_engine/video_capture_factory.h	Thu Nov 30 11:49:36 2023 +0000
@@ -23,7 +23,9 @@
  */
 class VideoCaptureFactory : webrtc::VideoCaptureOptions::Callback {
  public:
-  NS_INLINE_DECL_REFCOUNTING(VideoCaptureFactory);
+  NS_INLINE_DECL_THREADSAFE_REFCOUNTING(VideoCaptureFactory);
+
+  enum CameraAvailability { Unknown, Available, NotAvailable };
 
   VideoCaptureFactory();
 
@@ -48,6 +50,13 @@
    */
   RefPtr<CameraBackendInitPromise> InitCameraBackend();
 
+  /**
+   * Updates information about camera availability
+   */
+  using UpdateCameraAvailabilityPromise =
+      MozPromise<CameraAvailability, nsresult, true>;
+  RefPtr<UpdateCameraAvailabilityPromise> UpdateCameraAvailability();
+
  private:
   ~VideoCaptureFactory() = default;
   // aka OnCameraBackendInitialized
@@ -63,10 +72,11 @@
    *  2) NS_ERROR_NO_INTERFACE - the camera portal is not available
    *  3) NS_ERROR_UNEXPECTED - the camera portal returned wrong value
    */
-  using HasCameraDevicePromise = MozPromise<bool, nsresult, true>;
+  using HasCameraDevicePromise = MozPromise<CameraAvailability, nsresult, true>;
   RefPtr<HasCameraDevicePromise> HasCameraDevice();
 
   std::atomic<bool> mCameraBackendInitialized = false;
+  CameraAvailability mCameraAvailability = Unknown;
 #if (defined(WEBRTC_LINUX) || defined(WEBRTC_BSD)) && !defined(WEBRTC_ANDROID)
   std::unique_ptr<webrtc::VideoCaptureOptions> mVideoCaptureOptions;
 #endif